File: python/stack/src/stack.py
class Stack:
    def __init__(self, capacity=5):
        self.items = []
        self.capacity = capacity

    def is_empty(self):
        return len(self.items) == 0

    def is_full(self):
        return len(self.items) == self.capacity

    def push(self, data):
        if self.is_full():
            print("Stack overflow.")
        else:
            self.items.append(data)

    def pop(self):
        if self.is_empty():
            print("Stack underflow.")
            return None
        return self.items.pop()

    def peek(self):
        if self.is_empty():
            return None
        return self.items[-1]
    
    def print_stack(self):
        print("Stack:")
        for item in self.items:
            print(item, end=" ")
        print()

File: rust/stack/src/stack.rs
use std::fmt::Debug;

pub struct Stack<T> {
    items: Vec<T>,
    capacity: usize,
}

impl<T: Debug> Stack<T> {
    pub fn new(capacity: usize) -> Self {
        Stack {
            items: Vec::with_capacity(capacity),
            capacity,
        }
    }

    pub fn is_empty(&self) -> bool {
        self.items.is_empty()
    }

    pub fn is_full(&self) -> bool {
        self.items.len() == self.capacity
    }

    pub fn push(&mut self, data: T) {
        if self.is_full() {
            println!("Stack overflow!");
        } else {
            self.items.push(data);
        }
    }

    pub fn pop(&mut self) -> Option<T> {
        if self.is_empty() {
            println!("Stack underflow!");
            None
        } else {
            self.items.pop()
        }
    }

    pub fn peek(&self) -> Option<&T> {
        self.items.last()
    }

    pub fn print_stack(&self) {
        for item in &self.items {
            println!("{:?}", item);
        }
    }
}